models/user/external_login_user.go

```diff
@@ -160,12 +160,36 @@ func UpdateExternalUserByExternalID(ctx context.Context, external *ExternalLogin
 	return err
 }
 
+// EnsureLinkExternalToUser link the external user to the user
+func EnsureLinkExternalToUser(external *ExternalLoginUser) error {
+	has, err := db.Exist[ExternalLoginUser](db.DefaultContext, builder.Eq{
+		"external_id":     external.ExternalID,
+		"login_source_id": external.LoginSourceID,
+	})
+	if err != nil {
+		return err
+	}
+
+	if has {
+		_, err = db.GetEngine(db.DefaultContext).Where("external_id=? AND login_source_id=?", external.ExternalID, external.LoginSourceID).AllCols().Update(external)
+		return err
+	}
+
+	_, err = db.GetEngine(db.DefaultContext).Insert(external)
+	return err
+}
+
 // FindExternalUserOptions represents an options to find external users
 type FindExternalUserOptions struct {
 	db.ListOptions
-	Provider string
-	UserID   int64
-	OrderBy  string
+	Provider        string
+	UserID          int64
+	LoginSourceID   int64
+	HasRefreshToken bool
+	Expired         bool
+	OrderBy         string
+	Limit           int
```

Comment:
why do you have Limit and Start here ... this should be covered by 	db.ListOptions ?!?

Reply:
Thanks! It looks like I accidentally kept this in during merge conflict resolution. Fixed in https://github.com/go-gitea/gitea/pull/31572/commits/0327ff2bf11296f14fbf06d98810f0565af19843